keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+   - _Bound_: the controller finished processing the request and has bound the BucketAccess to the BucketAccessRequest
+   - _Released_: the originating Bucket has been deleted, signalling that the Bucket is ready for garbage collection.  This will occur on Greenfield Buckets once all requests referencing the Bucket are deleted.
+   - _Failed_: error and all retries have been exhausted.
+   - _Retrying_: set when a driver or Kubernetes error is encountered during provisioning operations indicating a retry loop.
+
+#### BucketAccessClass
+
+A cluster-scoped API providing a way for admins to specify policies that may be used to access buckets.  These are always applicable in Greenfield, where access is dynamically granted, and only sometimes applicable in Brownfield, where a user identity may already exist, but not yet have access to a bucket.  In this case, a BucketAccessRequest will still specify the BucketAccessClass in order to determine which actions it is granted to a defined bucket.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketAccessClass
+metadata: 
+  name:
+provisioner: [1]
+supportedProtocols: [2]
+policyActions: [3]
+  allow:
+  - "*"
+  deny:
+  - "*"
+parameters: [4]
+```
+
+1. `provisioner`: The name of the provisioner that `BucketAccess` instances should be managed by.
+1. `supportedProtocols`: protocols the associated object store supports.  Applied when matching Bucket to BucketClasses.  Admins may specify more than protocol when necessary.  `BucketAccessRequests.spec.protocol` will be checked against this array prior to provisioning.
+1. `policyActions`: a set of provisioner/platform defined policy actions to allow or deny a given user identity.
+1. `parameters`:   (Optional)  A map of string, string key values.  Allows admins to control user and access provisioning by setting provisioner key-values.
+
+
+
+---
+
+## COSI Architecture (Warning - Construction Area)
+
+### Topology
+
+![](./bucket-api-diagrams-topology.png)
+
+
+
+### Greenfield
+
+In an automated system where bucket lifecycles are managed by a controller:
+
+1. A BucketClass is defined to enable the provisioning of object store storage, with a specified configuration.
+2. A user will define a [Bucket](#bucket) with the [BucketClass](#bucketclass) specified.  
+3. A [Bucket](#Bucket) is instantiated and encapsulates all configuration information from the BucketClass and Bucket.  The Bucket will store all necessary information about the provisioned storage instance, including connection data.  Credentials and other sensitive information should not be stored in a Bucket. 
+4. A driver will then provision the object storage and return the connection information, to be written to the Bucket.
+
+
+
+
+#### Sharing Dynamically Created Buckets
+
+Once a [Bucket](#Bucket) is created, it is discoverable by other users in the cluster.  
+
+5. In order to access the Bucket, a user may create a [Bucket](#bucket) that specifies the Bucket by name.
```

Comment:
"...a user may create a [BucketRequest](#bucketrequest) that specifies the Bucket by name."